File: packages/core/src/utils/jpDetect.ts
/**
 * Return true if the given text contains at least a couple of characters from
 * the Hiragana, Katakana, or CJK Unified Ideographs blocks. We require at
 * least 2 to avoid false positives on stray kanji in latin text.
 */
export function looksJapanese(text: string): boolean {
  if (!text) return false
  // ぀-ゟ hiragana, ゠-ヿ katakana, 一-鿿 CJK
  const re = /[぀-ゟ゠-ヿ一-鿿]/g
  let count = 0
  while (re.exec(text)) {
    count++
    if (count >= 2) return true
  }
  return false
}

File: lib/services/import/jp_detect.dart
final _jpRe = RegExp(r'[぀-ゟ゠-ヿ一-鿿]');

/// Returns true when the text contains at least two characters from the
/// Hiragana / Katakana / CJK blocks. The minimum threshold avoids false
/// positives on stray kanji embedded in latin prose.
bool looksJapanese(String text) {
  if (text.isEmpty) return false;
  var count = 0;
  for (final _ in _jpRe.allMatches(text)) {
    count++;
    if (count >= 2) return true;
  }
  return false;
}
